### PR TITLE
A pull request for adding diffusion model for fluid data super-resolution (dfsr)

### PR DESCRIPTION
**Pull request by dlshu**
# Modulus Pull Request

## Description

This is a pull request for adding the diffusion model for fluid data super-resolution for modulus. More details on the added or modified files have been added to the CHANGELOG - Dule Shu

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,20 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased] - 2024-03-21
+
+### Added
```
Comment (mnabian) [inline CHANGELOG.md:11]: Please just add a single line here, mentioning that you have added this example to Modulus. This level of detail is not needed and adding a single item is sufficient.

File: modulus/metrics/diffusion/loss.py [hunk 2]
```diff
@@ -576,3 +576,233 @@ def __call__(self, net, img_clean, img_lr, labels=None, augment_pipe=None):
         loss = weight * ((D_yn - y) ** 2)
 
         return loss
+
+class VELoss_dfsr:
+    """
+    Loss function corresponding to the variance exploding (VE) formulation.
```
Comment (mnabian) [inline modulus/metrics/diffusion/loss.py:582]: Could you please add more detail? How is this different compared to the `VELoss`?

File: modulus/metrics/diffusion/loss.py [hunk 3]
```diff
@@ -576,3 +576,233 @@ def __call__(self, net, img_clean, img_lr, labels=None, augment_pipe=None):
...
+        num_diffusion_timesteps = b.shape[0]
+        
+        output = net(x, t, labels)
+        loss = (e - output).square()
+        
+        return loss
+
+class VELoss_dfsr_cond:
+    """
+    Loss function corresponding to the variance exploding (VE) formulation.
```
Comment (mnabian) [inline modulus/metrics/diffusion/loss.py:697]: Same here. Please add more details.

File: modulus/models/diffusion/preconditioning.py [hunk 4]
```diff
@@ -975,3 +975,257 @@ def __init__(
             sigma_max=sigma_max,
             sigma_data=sigma_data,
         )
+
+class VEPrecond_dfsr(torch.nn.Module):
+    """
+    Preconditioning corresponding to the variance exploding (VE) formulation.
```
Comment (mnabian) [inline modulus/models/diffusion/preconditioning.py:981]: Please add more details.

File: modulus/models/diffusion/preconditioning.py [hunk 5]
```diff
@@ -975,3 +975,257 @@ def __init__(
...
+        
+        if F_x.dtype != dtype:
+            raise ValueError(
+                f"Expected the dtype to be {dtype}, but got {F_x.dtype} instead."
+            )
+
+        return F_x
+
+
+class VEPrecond_dfsr_cond(torch.nn.Module):
```
Comment (mnabian) [inline modulus/models/diffusion/preconditioning.py:1072]: Please add more details

**Comment by mnabian** [comment]
@dlshu please add unit tests for your loss functions and preconditioners.

File: examples/generative/diffusion/train.py [hunk 6]
```diff
@@ -282,6 +311,18 @@ def main(cfg: DictConfig) -> None:
 # ----------------------------------------------------------------------------
 
 if __name__ == "__main__":
-    main()
+    parser = argparse.ArgumentParser()
```
Comment (mnabian) [inline examples/generative/diffusion/train.py:314]: You don't need to use argparse to get the config file, hydra already has support for this. Please check out the CorrDiff example to see how this is done: https://github.com/NVIDIA/modulus/blob/main/examples/generative/corrdiff/train.py#L48

File: examples/generative/diffusion/training_loop.py [hunk 7]
```diff
@@ -343,7 +351,15 @@ def training_loop(
 
         # Perform maintenance tasks once per tick.
         cur_nimg += batch_size
-        done = cur_nimg >= total_kimg * 1000
+        # done = cur_nimg >= total_kimg * 1000
+        if dataset == "dfsr":
+            done = cur_nimg >= total_kimg
+            if cur_nimg / batch_size % 500 == 0:
+                print("Progress in training iterations: loss: {}, iter: {}, cur_nimg: {}, \
```
Comment (mnabian) [inline examples/generative/diffusion/training_loop.py:358]: Use `logger0.info` instead of print statements. Example: https://github.com/NVIDIA/modulus/blob/main/examples/generative/corrdiff/train.py#L48

File: examples/generative/diffusion/training_loop.py [hunk 8]
```diff
@@ -410,6 +426,8 @@ def training_loop(
 
         # Save network snapshot.
         if (snapshot_ticks is not None) and (done or cur_tick % snapshot_ticks == 0):
+            if dataset == "dfsr":
+                print("Saving network snapshot.")
```
Comment (mnabian) [inline examples/generative/diffusion/training_loop.py:430]: Same here

File: examples/generative/diffusion/training_loop.py [hunk 9]
```diff
@@ -432,13 +450,19 @@ def training_loop(
...
-            and dist.rank == 0
-        ):
+        if dataset == "dfsr":
+            save_full_dump = (state_dump_ticks is not None) \
+                            and (done or cur_tick % state_dump_ticks == 0) \
+                            and dist.rank == 0
+        else:
+            save_full_dump = (state_dump_ticks is not None) \
+                            and (done or cur_tick % state_dump_ticks == 0) \
+                            and cur_tick != 0 \
```
Comment (mnabian) [inline examples/generative/diffusion/training_loop.py:460]: Let's just remove this and get rid of the entire conditional statement.

File: examples/generative/diffusion/training_loop.py [hunk 10]
```diff
@@ -432,13 +450,19 @@ def training_loop(
...
+                            and dist.rank == 0
+
+        if save_full_dump:
             # if (state_dump_ticks is not None) and (done or cur_tick % state_dump_ticks == 0) and dist.get_rank() == 0:
+            print("Saving full dump of the training state.")
```
Comment (mnabian) [inline examples/generative/diffusion/training_loop.py:465]: Use `logger0.info'

File: examples/generative/diffusion/training_loop.py [hunk 11]
```diff
@@ -470,5 +494,6 @@ def training_loop(
             break
 
     # Done.
-    logger0.info()
+    if not dataset == "dfsr":
```
Comment (mnabian) [inline examples/generative/diffusion/training_loop.py:497]: Is this needed? Maybe remove `logger0.info()` altogether

File: modulus/metrics/diffusion/loss.py (hunk 2 above)
Comment (dlshu) [reply modulus/metrics/diffusion/loss.py]: The comments in "class VELoss_dfsr" has been updated.

File: modulus/metrics/diffusion/loss.py (hunk 3 above)
Comment (dlshu) [reply modulus/metrics/diffusion/loss.py]: "class VELoss_dfsr_cond" has been removed due to its redundancy after "class VELoss_dfsr".

File: modulus/models/diffusion/preconditioning.py (hunk 5 above)
Comment (dlshu) [reply modulus/models/diffusion/preconditioning.py:1073]: The comments in "class VEPrecond_dfsr_cond" has been updated.

File: modulus/models/diffusion/preconditioning.py (hunk 4 above)
Comment (dlshu) [reply modulus/models/diffusion/preconditioning.py]: The comments in "class VEPrecond_dfsr" has been updated.

File: examples/generative/diffusion/train.py (hunk 6 above)
Comment (dlshu) [reply examples/generative/diffusion/train.py]: Thanks for sharing the reference from CorrDiff example. The code related to argparse has been removed. The command lines for model training and sampling have been updated in the README.md accordingly.

File: examples/generative/diffusion/training_loop.py (hunk 7 above)
Comment (dlshu) [reply examples/generative/diffusion/training_loop.py:358]: The ``print`` function has been replaced with ``logger0.info``.

File: examples/generative/diffusion/training_loop.py (hunk 10 above)
Comment (dlshu) [reply examples/generative/diffusion/training_loop.py:465]: The ``print`` function has been replaced with ``logger0.info``.

File: examples/generative/diffusion/training_loop.py (hunk 9 above)
Comment (dlshu) [reply examples/generative/diffusion/training_loop.py:460]: Sure. The conditional statement has now been removed.

File: examples/generative/diffusion/training_loop.py (hunk 11 above)
Comment (dlshu) [reply examples/generative/diffusion/training_loop.py:497]: The ``logger0.info()`` code was skipped because it used to cause syntax error on my computer. This might be due to version mismatch of PythonLogger on my computer. The code ``logger0.info()`` has now been removed.

File: examples/generative/diffusion/training_loop.py (hunk 8 above)
Comment (dlshu) [reply examples/generative/diffusion/training_loop.py]: The print function has been replaced with logger0.info.

**Review by hasethinvd — COMMENTED** [review]
Everything looks good

**Review by mnabian — APPROVED** [review]

**Comment by mnabian** [comment]
@dlshu formatting test failed. Please black format your code according to the instructions here: https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md

File: examples/generative/diffusion/conf/config_dfsr_cond_generate.yaml [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+hydra:
+  job:
+    chdir: True
+  run:
+    dir: ./outputs/
+
+
+# Main options
+outdir: ./results_cond_01       # Where to save the results
```
Comment (ktangsali) [inline examples/generative/diffusion/conf/config_dfsr_cond_generate.yaml:23]: nit: the comments can be better indented similar to some other occurrences in this Yaml.

File: examples/generative/diffusion/generate_dfsr.py [hunk 13]
```diff
@@ -0,0 +1,942 @@
...
+from modulus.distributed import DistributedManager
+from modulus.launch.logging import PythonLogger, RankZeroLoggingWrapper
+from utils import EasyDict, construct_class_by_name
+import copy
+import logging
+import matplotlib.pyplot as plt
+from einops import rearrange
+from mpl_toolkits.axes_grid1 import ImageGrid
+import math
+import pickle
```
Comment (ktangsali) [inline examples/generative/diffusion/generate_dfsr.py:38]: Generally we like to avoid pickle library due to its security concerns. Might be fine inside an example directory though...

File: examples/generative/diffusion/generate_dfsr.py
Comment (ktangsali) [inline examples/generative/diffusion/generate_dfsr.py:1]: nit: this seems like a very long training script. possibly the utils can be moved to a separate file...

File: test/metrics/diffusion/test_losses.py [hunk 14]
```diff
@@ -265,3 +266,46 @@ def mock_augment_pipe(imgs):
 #         fake_net, img_clean, img_lr, labels, mock_augment_pipe
 #     )
 #     assert isinstance(loss_value_with_augmentation, torch.Tensor)
+
+
+# VELoss_dfsr tests
+
+
+def test_veloss_dfsr_initialization():
```
Comment (ktangsali) [inline test/metrics/diffusion/test_losses.py:274]: no parameterization? cpu / gpu? A few other values?

File: modulus/metrics/diffusion/loss.py [hunk 15]
```diff
@@ -609,3 +610,131 @@ def __call__(self, net, img_clean, img_lr, labels=None, augment_pipe=None):
...
+        )
+        self.betas = torch.from_numpy(betas).float()
+        self.num_timesteps = betas.shape[0]
+
+    def get_beta_schedule(self, beta_schedule, *, beta_start, beta_end, num_diffusion_timesteps):
+        """
+        Compute the variance scheduling parameters {beta(0), ..., beta(t), ..., beta(T)} 
+        based on the VP formulation. 
+
+        beta_schedule: str
```
Comment (ktangsali) [inline modulus/metrics/diffusion/loss.py:654]: Would be good to document here the different methods available.

File: test/metrics/diffusion/test_losses.py [hunk 16]
```diff
@@ -265,3 +266,46 @@ def mock_augment_pipe(imgs):
...
+
+    loss_func = VELoss_dfsr(beta_start=0.0002, beta_end=0.01, num_diffusion_timesteps=500)
+    assert loss_func.beta_start == 0.0002
+    assert loss_func.beta_end == 0.01
+
+
+def test_get_beta_schedule_method():
+    loss_func = VPLoss_dfsr()
+
+    beta_schedule = 'linear'
```
Comment (ktangsali) [inline test/metrics/diffusion/test_losses.py:290]: probably test other methods?

**Comment by mnabian** [comment]
Thanks @ktangsali! Will make a note of these comments and fix in a separate PR.
